### PR TITLE
Par and request object only

### PR DESCRIPTION
**Pull request by tlodderstedt**
I added text to support PAR and request object only functionalities. 

Note: I added another server metadata parameter "require_request_objects" to allow the AS to indicate it generally requires request objects.

### CODE REVIEW

**Comment by panva** [comment]
> If the client uses dynamic client registration, setting the metadata parameter `request_object_signing_alg` MUST cause the authorization server to accept request objects for that particular client at the pushed authorization request endpoint only.

I don’t particularly agree with this. That option, as will be made clear by oidc errata has a meaning that shouldn’t be made different if the server supports par or not.

**Comment by tlodderstedt** [comment]
please propose text - the current text intentionally focuses on PAR but does not prescribe any behavior at another endpoint

**Comment by panva** [comment]
@tlodderstedt i believe you've made a typo in the PR when it comes that particular paragraph i quoted. I'll push to the same branch in a bit.

**Comment by panva** [comment]
@tlodderstedt i see the intent but the metadata `require_request_objects` being only applied to PAR or in general defining new metadata that is named generically but only applied to PAR just doesn't make sense to me.

Here's what i imagine the effects of each client/as metadata in isolation should be, WDYT? If you agree i can propose changes.

Screenshot 2020-05-10 20 52 09

<details>
<summary>Source markdown</summary>

```

| AS require_pushed_authorization_requests | AS require_request_objects | client request_object_signing_alg | client require_pushed_authorization_requests | Regular Authorization Request | Require Request Object | PAR |
|-|-|-|-|-|-|-|
| false | false | undefined | false | MAY | MAY (any alg) | MAY |
| **true** | false | undefined | false | **FORBIDDEN** | MAY (any alg) | **MUST** |
| false | **true** | undefined | false | MAY | **MUST** (any alg) | MAY |
| false | false | e.g. **PS256** | false | MAY | **MUST** (PS256) | MAY |
| false | false | undefined | **true** | **FORBIDDEN** | MAY (any alg) | **MUST** |

```
</details>

**Review by panva — CHANGES_REQUESTED** [review]
@tlodderstedt i see the intent but the metadata `require_request_objects` being only applied to PAR or in general defining new metadata that is named generically but only applied to PAR just doesn't make sense to me.

Here's what i imagine the effects of each client/as metadata in isolation should be, WDYT? If you agree i can propose changes.

Screenshot 2020-05-10 20 52 09

<details>
<summary>Source markdown</summary>

```

| AS require_pushed_authorization_requests | AS require_request_objects | client request_object_signing_alg | client require_pushed_authorization_requests | Regular Authorization Request | Require Request Object | PAR |
|-|-|-|-|-|-|-|
| false | false | undefined | false | MAY | MAY (any alg) | MAY |
| **true** | false | undefined | false | **FORBIDDEN** | MAY (any alg) | **MUST** |
| false | **true** | undefined | false | MAY | **MUST** (any alg) | MAY |
| false | false | e.g. **PS256** | false | MAY | **MUST** (PS256) | MAY |
| false | false | undefined | **true** | **FORBIDDEN** | MAY (any alg) | **MUST** |

```
</details>

**Comment by vdzhuvinov** [comment]
Is there a chance of having `require_request_objects` defined in the JAR spec?

**Comment by tlodderstedt** [comment]
@panva your proposal looks good. Can you please modify the text?
@vdzhuvinov given the adapted meaning, this makes sense. I suggest to reach out to the JAR authors once we have the text settled.

**Review by b---c — CHANGES_REQUESTED** [review]
Based on the proposal and discussion around this at https://mailarchive.ietf.org/arch/msg/oauth/qvfhZIz08yRtuEbgA7Ryka0pPno/ and that #43 was to capture and track that, I would have expected to see `require_pushed_authorization_requests` metadata and associated behavior defined for both client and AS in the body of the document. The IANA stuff is for registration and has a description that will show up in the registry(s) to help understand it. But the IANA section doesn't define the parameters - that's done in the document.  

The inclusion of require_request_objects in this change was unexpected and I don't think it is appropriate here.

**Comment by b---c** [comment]
also agree that PAR shouldn't be trying to apply special meaning to request_object_signing_alg

File: main.md [hunk 1]
```diff
@@ -302,11 +311,12 @@ The client uses the `request_uri` value returned by the authorization server to
   request_uri=urn%3Aexample%3Abwc4JK-ESC0w8acc191e-Y1LTC2 HTTP/1.1
 ```
 
-# Authorization Server Metadata
+# Error Authorization Response  
+An authorization server MAY decide to support pushed authorization requests as the only means for clients to pass authorization request data. In this case, the authorization server will refuse to process any authorization request not fulfilling the requirements as define in (#authz_request) and return an error `invalid_request`.
```
Comment (b---c) [inline main.md:315]: maybe unauthorized_client ?

File: main.md [hunk 2]
```diff
@@ -232,6 +240,8 @@ For an error the authorization server sets an appropriate HTTP status code and M
 
 If the authorization server sets an error code, it SHOULD be one of the defined codes for the token endpoint in Section 5.2 or for the authorization endpoint in Sections 4.1.2.1 and 4.2.2.1 of [@!RFC6749], or by an OAuth extension if one is involved in the initial processing of authorization request that was pushed. Since initial processing of the pushed authorisation request doesn't involve resource owner interaction, error codes related to user interaction, such as `consent_required` defined by [@!OIDC], are not returned.
 
+If the authorization server or the client's policy require use of request objects only at the pushed authorization endpoint, the authorization server will return a HTTP status code 400 along with the error `invalid_request`.
```
Comment (b---c) [inline main.md:243]: unauthorized_client ?

File: main.md [hunk 3]
```diff
@@ -147,7 +147,15 @@ This specification uses the terms "access token", "refresh token",
...
+
+The endpoint accepts the parameters defined in [@!RFC6749] for the authorization endpoint as well as all applicable extensions defined for the authorization endpoint. Some examples of such extensions include PKCE [@RFC7636], Resource Indicators [@RFC8707], and OpenID Connect [@OIDC]. The endpoint also supports sending all authorization request parameters as request object according to [@!I-D.ietf-oauth-jwsreq].
+
+Note: the authorization server or the client MAY require to use request objects only. 
+
+If the authorization server requires request objects at its pushed authorization endpoint, it SHOULD include the `require_request_objects` Server Metadata parameter as defined in (#server_metadata) in its discovery responses.
+
+If the client uses dynamic client registration, setting the metadata parameter `request_object_signing_alg` MUST cause the authorization server to accept request objects for that particular client at the pushed authorization request endpoint only.
```
Comment (b---c) [inline main.md:158]: don't think this should be here

**Comment by tlodderstedt** [comment]
@b---c 

> The inclusion of require_request_objects in this change was unexpected and I don't think it is appropriate here.

Why do you consider this inappropriate? Don't you think the AS should be able to decide to only accept request objects?

**Comment by b---c** [comment]
> Why do you consider this inappropriate? 

Because there was some discussion and agreement around adding require_pushed_authorization_requests for client and server on the WG list for this document that is now a WG document. So that was the change I would have expected to see. And was even planning on making. But then that got put into a change with other stuff.

And require_request_objects was just thrown onto this PR. And that complicates reviewing and approving text for a change that was otherwise agreed to. 

Also, the name require_request_objects is general and independent of PAR so shouldn't have PAR specific meaning. And if it's not PAR specific but applicable to JAR in general, then it really shouldn't be defined in PAR but should be in JAR. Though I recognize that getting changes into JAR is complicated in its own right.  

> Don't you think the AS should be able to decide to only accept request objects?

An AS can decide that, of course. I'm not sure that a metadata parameter is needed for it though. And that should be a different conversation.

**Comment by tlodderstedt** [comment]
I checked the respective threat again. 

https://mailarchive.ietf.org/arch/msg/oauth/1mzeKcb7I78AOGL8t-qNukkZyYc/

I had raised the question whether the AS should be able to require request object but the discussion then focussed on adoption of the OIDC mechanism to determine the client specific policy. When I wrote the PR, I remembered and added this metadata parameter since I thought it would be easier to discuss a concrete proposal. 

Sorry for any confusing this has caused. I will bring this topic to the list again.

**Comment by panva** [comment]
Feel free to use the graphic i posted above. Imho the client alg property serves the same purpose and the new one is redundant.

**Comment by tlodderstedt** [comment]
@panva Why does it serve the same purpose? In my opinion, it's a difference whether clients register with the intend to use request objects only or the AS deciding to only accept request objects.

**Comment by panva** [comment]
> @panva Why does it serve the same purpose? In my opinion, it's a difference whether clients register with the intend to use request objects only or the AS deciding to only accept request objects.

fair enough.

**Comment by panva** [comment]
On today's OIDC WG Call the presence of `request_object_signing_alg` as a signal to require the use of request objects was discussed and kind of disputed again with unclear way forward - see https://bitbucket.org/openid/connect/issues/1171/creating-a-way-to-mandate-request-object and https://bitbucket.org/openid/connect/issues/1045/signalling-that-a-request-object-must

**Pull request closed without merging**